lib/remap-lang/src/function.rs

```diff
@@ -0,0 +1,124 @@
+use crate::{Expr, Expression, Result, Value};
+use core::convert::TryInto;
+use std::collections::HashMap;
+
+mod split;
+
+pub(crate) use split::Split;
+
+#[derive(thiserror::Error, Debug, PartialEq)]
+pub enum Error {
+    #[error(r#"expected expression argument, got regex"#)]
+    ArgumentExprRegex,
+
+    #[error(r#"missing required argument "{0}""#)]
+    Required(String),
+}
+
+#[derive(Copy, Clone)]
+pub(crate) struct Parameter {
+    /// The keyword of the parameter.
+    ///
+    /// Arguments can be passed in both using the keyword, or as a positional
+    /// argument.
+    pub keyword: &'static str,
+
+    /// The parser calls this method to determine if a given argument value is
+    /// accepted by the parameter.
+    pub accepts: fn(&Value) -> bool,
+
+    /// Whether or not this is a required parameter.
+    ///
+    /// If it isn't, the function can be called without errors, even if the
+    /// argument matching this parameter is missing.
+    pub required: bool,
+}
+
+impl std::fmt::Debug for Parameter {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Parameter")
+            .field("keyword", &self.keyword)
+            .field("required", &self.required)
+            .field("accepts", &"fn(&Value) -> bool".to_owned())
+            .finish()
+    }
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct ArgumentList(HashMap<&'static str, Argument>);
+
+impl ArgumentList {
+    pub fn optional(&mut self, keyword: &str) -> Option<Argument> {
+        self.0.remove(keyword)
+    }
+
+    pub fn optional_expr(&mut self, keyword: &str) -> Result<Option<Box<Expr>>> {
+        self.optional(keyword)
+            .map(|v| v.try_into().map(Box::new).map_err(Into::into))
+            .transpose()
+    }
+
+    pub fn required(&mut self, keyword: &str) -> Result<Argument> {
+        self.0
+            .remove(keyword)
+            .ok_or_else(|| Error::Required(keyword.to_owned()).into())
+    }
+
+    pub fn required_expr(&mut self, keyword: &str) -> Result<Box<Expr>> {
+        self.required(keyword)
+            .and_then(|v| v.try_into().map_err(Into::into))
+            .map(Box::new)
+    }
+
+    pub fn keywords(&self) -> Vec<&'static str> {
+        self.0.keys().copied().collect::<Vec<_>>()
+    }
+
+    pub fn insert(&mut self, k: &'static str, v: Argument) {
+        self.0.insert(k, v);
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum Argument {
+    Expression(Expr),
+    Regex(regex::Regex),
+}
+
+impl TryInto<Expr> for Argument {
+    type Error = Error;
+
+    fn try_into(self) -> std::result::Result<Expr, Self::Error> {
+        match self {
+            Argument::Expression(expr) => Ok(expr),
+            Argument::Regex(_) => Err(Error::ArgumentExprRegex),
+        }
+    }
+}
+
+pub(crate) trait Function: std::fmt::Debug {
```

Comment:
It is confusing having both a trait and a struct called `Function`. Could this be renamed?

Reply:
I'd rather not. This trait is publicly accessible (it isn't in this PR, but it is made public in #4709), and I think the name make sense.

Similarly, the name `expression::Function` also makes sense.

They are in different modules, and outside this crate, you only deal with one of the two (the trait), so I don't think we need to change anything here.